himbaechel/uarch/ng-ultra/ng_ultra.cc

```diff
@@ -0,0 +1,1027 @@
+                            phys_to_log[extra_data.input].push_back(i);
+                        }
+                    }
+                }
+                unsigned permuted_init = 0;
+                for (unsigned i = 0; i < 16; i++) {
+                    unsigned log_idx = 0;
+                    for (unsigned j = 0; j < 4; j++) {
+                        if ((i >> j) & 0x1) {
+                            for (auto log_pin : phys_to_log[j])
+                                log_idx |= (1 << log_pin);
+                        }
+                    }
+                    if ((orig_init >> log_idx) & 0x1)
+                        permuted_init |= (1 << i);
+                }
+                cell.second->params[id_lut_table] = Property(permuted_init, 16);
+            }
+        }
+    }
+
+    fixup_crossbars();
+
+    print_utilisation(ctx);
+    const ArchArgs &args = ctx->args;
+    if (args.options.count("bit")) {
+        write_bitstream_json(args.options.at("bit"));
+    }
+}
+
+void NgUltraImpl::configurePlacerHeap(PlacerHeapCfg &cfg)
+{
+    cfg.beta = 0.5;
+    cfg.placeAllAtOnce = true;
+}
+
+namespace {
+
+template <size_t N> bool check_assign_sig(std::array<const NetInfo*, N> &sig_set, const NetInfo *sig)
+{
+    if (sig == nullptr)
+        return true;
+    for (size_t i = 0; i < N; i++)
+        if (sig_set[i] == sig) {
+            return true;
+        } else if (sig_set[i] == nullptr) {
+            sig_set[i] = sig;
+            return true;
+        }
+    return false;
+};
+
+struct SectionFEWorker
+{
+    std::array<const NetInfo *, 2> clk{}; // from local system matrix
+    std::array<const NetInfo *, 2> reset{}; // from local system matrix
+    std::array<const NetInfo *, 2> load{}; // from local system matrix
+    std::array<const NetInfo *, 1> shared{}; // 1 from local system matrix
+    // Additional R and L can be used from RI network
+    bool run(const NgUltraImpl *impl,const Context *ctx, BelId bel, CellInfo *cell)
+    {
+        Loc loc = ctx->getBelLocation(bel);
+        for (uint8_t id = 0; id <= BEL_LUT_MAX_Z; id++) {
+            const CellInfo *ff = ctx->getBoundBelCell(ctx->getBelByLocation(Loc(loc.x,loc.y,id)));
+            if (ff == nullptr)
+                continue;
+            if (!check_assign_sig(reset, ff->getPort(id_R))) {
+                if (!check_assign_sig(shared, ff->getPort(id_R)))
+                    return false;
+            }
+            if (!check_assign_sig(load, ff->getPort(id_L))) {
+                if (!check_assign_sig(shared, ff->getPort(id_L)))
+                    return false;
+            }
+            if (!check_assign_sig(clk, ff->getPort(id_CK)))
+                return false;
+        }
+        const auto &bel_data = chip_bel_info(ctx->chip_info, bel);
+        const auto &extra_data = *reinterpret_cast<const NGUltraBelExtraDataPOD *>(bel_data.extra_data.get());
+        std::string type = str_or_default(cell->params, ctx->id("type"), "");
```

Comment:
let's make this faster than copying a string, this could well be in the critical path

Reply:
Something like this I can do, believe other approaches would require refactoring larger amount of code, and I do not have hardware to work on any more.

```
        if (cell->params.count(id_type)) {
            const std::string &type = cell->params[id_type].as_string();
            if (type=="CSC" && (extra_data.flags & BEL_EXTRA_FE_CSC) == 0) return false; // No CSC capability on FE
            if (type=="SCC" && (extra_data.flags & BEL_EXTRA_FE_SCC) == 0) return false; // No SCC capability on FE
        }

```

Reply:
yeah, better than nothing